sky/execution.py

```diff
@@ -943,3 +956,319 @@ def _maybe_translate_local_file_mounts_and_sync_up(task: task_lib.Task):
                     raise exceptions.NotSupportedError(
                         f'Unsupported store type: {store_type}')
             storage_obj.force_delete = True
+
+
+@usage_lib.entrypoint
+def serve_up(
+    task: 'sky.Task',
+    service_name: str,
+    controller_resources: 'sky.Resources',
+    controller_best_resources: 'sky.Resources',
+):
+    """Spin up a service.
+
+    Please refer to the sky.cli.serve_up for the document.
+
+    Args:
+        task: sky.Task to serve up.
+        service_name: Name of the service.
+        controller_resources: The resources requirement for the controller.
+        controller_best_resources: The optimized resources for the controller.
+    """
+    controller_cluster_name = serve.generate_controller_cluster_name(
+        service_name)
+    controller_best_resources.cloud.check_cluster_name_is_valid(
+        controller_cluster_name)
+    assert task.service is not None, task
+    assert len(task.resources) == 1, task
+    requested_resources = list(task.resources)[0]
+    service_handle = serve.ServiceHandle(
+        controller_cluster_name=controller_cluster_name,
+        policy=task.service.policy_str(),
+        requested_resources=requested_resources,
+        replica_info=[])
+    global_user_state.add_or_update_service(
+        service_name, None, service_handle,
+        status_lib.ServiceStatus.CONTROLLER_INIT)
+    app_port = int(task.service.app_port)
+
+    # TODO(tian): Use skyserve constants.
+    _maybe_translate_local_file_mounts_and_sync_up(task)
+    ephemeral_storage = []
+    if task.storage_mounts is not None:
+        for storage in task.storage_mounts.values():
+            if not storage.persistent:
+                ephemeral_storage.append(storage.to_yaml_config())
+    service_handle.ephemeral_storage = ephemeral_storage
+    global_user_state.set_service_handle(service_name, service_handle)
+
+    with tempfile.NamedTemporaryFile(prefix=f'serve-task-{service_name}-',
+                                     mode='w') as f:
+        task_config = task.to_yaml_config()
+        if 'resources' in task_config and 'spot_recovery' in task_config[
+                'resources']:
+            del task_config['resources']['spot_recovery']
```

Comment:
This is due to how the `Resources` class operates on this variable... If this field is None, instead of other fields, there will be a `'spot_recovery': None` in the yaml config which is causing some parsing error for the generated task yaml. Was this intended? IIUC it is due to this line not using `add_if_not_none('spot_recovery', self.spot_recovery)`, but `config['spot_recovery'] = self.spot_recovery`.
https://github.com/skypilot-org/skypilot/blob/bea2fc2aa1ae7a896792231d85d285b5312d436f/sky/resources.py#L1137C53-L1137C53

Reply:
A test:

```python
Python 3.9.17 (main, Jul  5 2023, 20:41:20) 
[GCC 11.2.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sky
>>> sky.Resources().to_yaml_config()
{'spot_recovery': None, 'disk_size': 256}
```

Reply:
After an offline discussion, we decided to remove `spot_recovery` if it is None